lib/build_string.rb

```diff
@@ -1,7 +1,136 @@
+  end
+
+  # given a rhyme, which is a string, returns array of elements. 
+  # Ex: "1- the house that Jack built, 2- the malt that lay in, etc"
+  def get_elements
+    lines = @rhyme.each_line.to_a
+    elements = []
+    # number of words in previous element
+    prev_length = 0
+
+    lines.each do |line|
+      words = line.split 
+      diff = words.length - prev_length
+    
+      string, prev_length = get_element_in_line(diff, words, prev_length)
+
+      elements.push(string)
+    end
+    elements
+  end
+  
+  # returns "new" element within 1 line of rhyme, 
+  # i.e. excludes "This is" and repeated elements from previous lines.
+  def get_element_in_line(diff, words, prev_length)
+    element = []
+    (diff).times do |i|
+      next if words[i] == "This" || words[i]== "is"
+      # concatenate word into element string
+      element << words[i]
+    end
+    prev_length += element.length
+    string = element.join(" ")
+    
+    if string[-1] == "."
+      string = string.chop
+    end
+    return [string, prev_length]
+  end
+
+  # given a rhyme as a string, return a semi-random version
+  # that always ends with "the house that Jack built".
+  def randomize_rhyme
+    elements = get_elements
+    # shuffle method does not modify the original array in place, it returns a modified version.
+    shuffled_elements = elements.shuffle 
+
+    # return the resulting string of rhyme with randomized order of elements,
+    # excluding the final newline char
+    concat_elements(shuffled_elements)
+  end
+
+  # given a rhyme as a string, return randomized version.
+  def randomize_rhyme
+    elements = get_elements
+    # shuffle method does not modify the original array in place, it returns a modified version.
+    shuffled_elements = elements.shuffle 
+
+    # return the resulting string of rhyme with randomized order of elements,
+    # excluding the final newline char
+    concat_elements(shuffled_elements)
+  end
+
+  # helper method of randomize_rhyme to maintain pattern despite being random
+  def concat_elements(shuffled_elements)
+    output = ""
+    saved = ""
+    shuffled_elements.each do |element|
+      if saved == ""
+        output += "This is #{element}.\n"
+        saved = "#{element}"
+      else
+        output += "This is #{element} #{saved}.\n "
+        saved = "#{element} #{saved}"
+      end
+    end
+    output.chop
+  end
+
+  # pass a string (rhyme) and confirm that it has correct pattern
+  def validate_rhyme(string)
+    lines = string.each_line.to_a
+    saved = []
+    saved_length = 0
+
+    lines.each do |line|
+      words = line.split
+      line_length = words.length
+      diff = line_length - saved_length
+
+      check = validate_first_words(words)
+      return false if check == false
+
+      check = validate_last_part(words, saved, diff, line_length)
+      return false if check == false
+
+      saved = validate_save_middle_part(line_length, words)
```

Comment:
Why is this 'saved' after you validate the [validate_last_part(words, saved, diff, line_length)](https://github.com/pulibrary/Ruby-Checkpoint-Liz/pull/9/files#diff-6542a64ac46ae69f8b48bc227ddd2499149fe03288b8c18e8732c777cd567e81R97)

Reply:
the validate_save_middle_part is a helper function of the validate_rhyme function. This helper method returns the middle part of the current line that is to be saved for the next iteration since we need to know the length of the saved string to calculate diff.